pkg/server/delta/v3/server.go

```diff
@@ -44,6 +132,340 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
+				state.ResourceVersions[r] = v
+			}
+			if unsub := req.GetResourceNamesUnsubscribe(); len(unsub) > 0 {
+				s.unsubscribe(unsub, state.ResourceVersions)
+			}
+
+			if s.callbacks != nil {
+				if err := s.callbacks.OnStreamDeltaRequest(streamID, req); err != nil {
+					return err
+				}
+			}
+
+			// cancel existing watches to (re-)request a newer version
+			switch {
+			case req.TypeUrl == resource.EndpointType:
+				if values.deltaEndpointNonce == "" || values.deltaEndpointNonce == nonce {
+					if values.deltaEndpointCancel != nil {
+						values.deltaEndpointCancel()
+					}
+					values.deltaEndpoints, values.deltaEndpointCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.ClusterType:
+				if values.deltaClusterNonce == "" || values.deltaClusterNonce == nonce {
+					if values.deltaClusterCancel != nil {
+						values.deltaClusterCancel()
+					}
+					values.deltaClusters, values.deltaClusterCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.RouteType:
+				if values.deltaRouteNonce == "" || values.deltaRouteNonce == nonce {
+					if values.deltaRouteCancel != nil {
+						values.deltaRouteCancel()
+					}
+					values.deltaRoutes, values.deltaRouteCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.ListenerType:
+				if values.deltaListenerNonce == "" || values.deltaListenerNonce == nonce {
+					if values.deltaListenerCancel != nil {
+						values.deltaListenerCancel()
+					}
+					values.deltaListeners, values.deltaListenerCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.SecretType:
+				if values.deltaSecretNonce == "" || values.deltaSecretNonce == nonce {
+					if values.deltaSecretCancel != nil {
+						values.deltaSecretCancel()
+					}
+					values.deltaSecrets, values.deltaSecretCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.RuntimeType:
+				if values.deltaRuntimeNonce == "" || values.deltaRuntimeNonce == nonce {
+					if values.deltaRuntimeCancel != nil {
+						values.deltaRuntimeCancel()
+					}
+					values.deltaRuntimes, values.deltaRuntimeCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			default:
+				typeURL := req.TypeUrl
+				responseNonce, seen := values.deltaNonces[typeURL]
+				if !seen || responseNonce == nonce {
+					// We must signal goroutine termination to prevent a race between the cancel closing the watch
+					// and the producer closing the watch.
+					if terminate, exists := values.deltaTerminations[typeURL]; exists {
+						close(terminate)
+					}
+					if cancel, seen := values.deltaCancellations[typeURL]; seen && cancel != nil {
+						cancel()
+					}
+
+					var watch chan cache.DeltaResponse
+					watch, values.deltaCancellations[typeURL] = s.cache.CreateDeltaWatch(req, &state)
+
+					// a go-routine. Golang does not allow selecting over a dynamic set of channels.
```

Comment:
This is not really true (https://github.com/envoyproxy/go-control-plane/issues/413) but I think it's okay to land this as is (since it matches sotw) and iterate as we get some throughput benchmarks set up

Reply:
Sounds great, that's on my list to do after we land all this. I saw that issue I'll brush up on it again